drivers/sensor/pixart/pat9136/pat9136_stream.c

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) 2025 Croxel Inc.
+ * Copyright (c) 2025 CogniPilot Foundation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/sensor_clock.h>
+#include <zephyr/sys/check.h>
+#include <zephyr/rtio/rtio.h>
+#include <zephyr/rtio/work.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/drivers/sensor/pat9136.h>
+
+#include "pat9136.h"
+#include "pat9136_reg.h"
+#include "pat9136_stream.h"
+#include "pat9136_decoder.h"
+#include "pat9136_bus.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(PAT9136_STREAM, CONFIG_SENSOR_LOG_LEVEL);
+
+static void start_drdy_backup_timer(const struct device *dev)
+{
+	struct pat9136_data *data = dev->data;
+	const struct pat9136_config *cfg = dev->config;
+
+	k_timer_start(&data->stream.timer.backup,
+		      K_MSEC(cfg->backup_timer_period),
+		      K_NO_WAIT);
+}
+
+static void start_drdy_cooldown_timer(const struct device *dev)
+{
+	struct pat9136_data *data = dev->data;
+	const struct pat9136_config *cfg = dev->config;
+
+	k_timer_start(&data->stream.timer.cooldown.timer,
+		      K_MSEC(cfg->cooldown_timer_period),
+		      K_NO_WAIT);
+}
+
+static void pat9136_complete_result(struct rtio *ctx,
+				    const struct rtio_sqe *sqe,
+				    void *arg)
+{
+	const struct device *dev = (const struct device *)arg;
+	struct pat9136_data *data = dev->data;
+	struct rtio_iodev_sqe *iodev_sqe = data->stream.iodev_sqe;
+	struct pat9136_encoded_data *edata = sqe->userdata;
+
+	data->stream.iodev_sqe = NULL;
+
+	edata->header.events.drdy = true &&
+				    data->stream.settings.enabled.drdy;
+	edata->header.events.motion = REG_MOTION_DETECTED(edata->motion) &&
+				      data->stream.settings.enabled.motion;
+	edata->header.channels = 0;
+
+	if ((data->stream.settings.enabled.drdy &&
+	      data->stream.settings.opt.drdy == SENSOR_STREAM_DATA_INCLUDE) ||
+	    (data->stream.settings.enabled.motion &&
+	      data->stream.settings.opt.motion == SENSOR_STREAM_DATA_INCLUDE)) {
+		edata->header.channels |= pat9136_encode_channel(SENSOR_CHAN_POS_DXYZ);
+		edata->header.channels |= pat9136_encode_channel(SENSOR_CHAN_POS_DXYZ_MM);
+	}
+
+	if (data->stream.settings.enabled.drdy) {
+		start_drdy_backup_timer(dev);
+	}
+
+	/** Attempt chip recovery if erratic behavior is detected  */
+	if (!REG_OBSERVATION_READ_IS_VALID(edata->observation)) {
+
+		LOG_WRN("CHIP OK register indicates issues. Attempting chip recovery: 0x%02X",
+			edata->observation);
+
+		rtio_iodev_sqe_err(iodev_sqe, -EAGAIN);
+	} else {
+		rtio_iodev_sqe_ok(iodev_sqe, 0);
+	}
+}
+
+static void pat9136_stream_get_data(const struct device *dev)
+{
+	struct pat9136_data *data = dev->data;
+	uint64_t cycles;
+	int err;
+
+	CHECKIF(!data->stream.iodev_sqe) {
+		LOG_WRN("No RTIO submission with an INT GPIO event");
+		return;
+	}
+
+	start_drdy_cooldown_timer(dev);
```

Comment:
It seems like the cooldown timer is here to rate limit interrupt handling so the mcu isn't DoSed by interrupts is that right? I think this deserves some comments on how this works in the file so people who read the code might learn from what you've done easier.

Reply:
That's correct. I've added comments describing its purpose and how it works. Hopefully this is enough.